### ISSUE TITLE
Support non-standard cljc reader conditionals

### ISSUE THREAD

**Issue by tangrammer** [clj-kondo/clj-kondo#1686]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

<hr>

EDITED by @borkdude: 

We should have a more general solution than supporting a fixed number of non-standard cljc reader conditionals.
The only supported conditionals right now are `:clj` and `:cljs`. We should however support an open world. 

```
:cljc {:features [:node :browser :clj]}
```

and also have a way to specify which reader conditional is similar to a supported one:

```
:cljc {:lint-feature-as {:node :cljs}}
```

<hr>

**Is your feature request related to a problem? Please describe.**
My project is a shadow-cljs one, that targets to `:node`, `:browser` and `:clj` 
There we are using normal approach to [require namespaces](https://clojure.org/guides/reader_conditionals#_namespaces), example [code](https://github.com/kubelt/kubelt/blob/1f08518395c681db4b4c6cb1adb1f181179424d5/src/main/com/kubelt/ipfs/client.cljc#L14-L22) 
shadow-cljs works fine but using clj-kondo throw warnings when we use the specific namespace, example [code](https://github.com/kubelt/kubelt/blob/1f08518395c681db4b4c6cb1adb1f181179424d5/src/main/com/kubelt/ipfs/client.cljc#L250)

This is the warning output
```
------ WARNING #1 - :undeclared-ns ---------------------------------------------
 File: /Users/tangrammer/git/kubelt/kubelt/src/main/com/kubelt/ipfs/client.cljc:326:35
--------------------------------------------------------------------------------
 323 |              (if (contains? options :http/client)
 324 |                (let [http-client (get options :http/client)]
 325 |                  {:http/client http-client})
 326 |                (let [http-client (lib.http/->HttpClient)]
-----------------------------------------^--------------------------------------
 No such namespace: lib.http, could not locate lib/http.cljs, lib/http.cljc, or JavaScript source providing "lib.http"
--------------------------------------------------------------------------------
 327 |                  {:http/client http-client}))]
 328 |          (if-not (satisfies? proto.http/HttpClient client)
 329 |            {:com.kubelt/type :kubelt.type/error
 330 |             :error "invalid HTTP client"}
--------------------------------------------------------------------------------
```

**Describe the solution you'd like**
Support shadow-cljs conditional readers

**Describe alternatives you've considered**
don't know any

Perhaps I could contribute to the fix if i get a bit of guiding :)

**Comment by borkdude** [comment]
This relates to #168 #2072

**Comment by markokocic** [comment]
Just came across this. It would be very useful to have the ability to lint code related to different dialicts.

For example, I work on an app using Babashka, and now I'm trying to add Jolt support. I have a need for #?(jolt and bb) reader conditionals, but clj-kondo doesn't support it.

Both Jolt and Babashka work quite well with those conditionals, but clj-kondo reports errors since it doesn't have the ability to lint them as standard Clojure using Clojure rules.

**Comment by borkdude** [comment]
As a temporary workaround you can add this to you configuration: `:cljc {:features #{:clj}}` . then it will only lint the `:clj` branch and not warn about a missing clojurescript branch. I understand that this isn't fully supporting bb, jolt and other dialects yet, but that's coming. The best way to raise priority is to upvote with a thumbs up on the original message.

**Comment by markokocic** [comment]
Thanks @borkdude , that's what I already did. The issue is, the app is currently babashka only, with the ambition to add Jolt as a target in the near future. Using :clj in reader instead of :bb and :jolt just to make clj-kondo to check it feels a bit dishonest :(

**Comment by borkdude** [comment]
@markokocic if the app is for babashka only, then probably `:clj` is justified since babashka is intended to be compatible with :clj? and why not make your app compatible with :clj as well?

**Comment by markokocic** [comment]
You are right. But so is Jolt. Both are Clojure dialects, both have :clj in feature flags, but still different.

The main reason for lack of Clojure is I want fast startup time and native distribution, so I restricted it to Babashka and its ecosystem. I already paid the price of avoiding JVM dependencies. Theoretically, it may work with Clojure with a few tweaks, but I just didn't bother to do it.

I opted for Jolt as a second target instead because it shares a lot of ideas with Babashka, and out of pure curiosity. I wanted to see how far I could get it to work for both platforms.
